src/main/java/org/javarosa/xml/InternalDataInstanceParser.java

```diff
@@ -0,0 +1,64 @@
+package org.javarosa.xml;
+
+import org.javarosa.core.model.instance.DataInstance;
+import org.javarosa.core.model.instance.FormInstance;
+import org.javarosa.core.model.instance.TreeElement;
+import org.javarosa.core.reference.InvalidReferenceException;
+import org.javarosa.xml.util.InvalidStructureException;
+import org.javarosa.xml.util.UnfullfilledRequirementsException;
+import org.kxml2.io.KXmlParser;
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * @author johnthebeloved
+ * Parses the internal secondary instances out from an Xform file
+ */
+public class InternalDataInstanceParser {
+
+
+    /**
+     * Builds all of the <b>InternalDataInstances</b> contained of an
+     * XForm
+     *
+     * Implemented as an alternative to parsing the each internal
+     * instance
+     *
+     * @param xFormSrc the path  of the Xform containing internal instances…
+     * @return List of all the #InternalDataInstance in the XForm
+     * @throws IOException                       if FileInputStream can’t find the file, or ElementParser can’t read the stream
+     * @throws InvalidReferenceException         if the ReferenceManager in getPath(String srcLocation) can’t derive a reference
+     * @throws UnfullfilledRequirementsException thrown by {@link TreeElementParser#parse()}
+     * @throws XmlPullParserException            thrown by {@link TreeElementParser#parse()}
+     * @throws InvalidStructureException         thrown by {@link TreeElementParser#parse()}
+     */
+    public static HashMap<String, DataInstance> buildInstances(String xFormSrc)
+        {
+//throws IOException, UnfullfilledRequirementsException, XmlPullParserException, InvalidStructureException, InvalidReferenceException
```

Comment:
The fact that these exceptions are caught below means we could possibly end up in a state where JR thinks it has read in all secondary instances when deserializing a form but in fact some or all have failed, right? I believe that the user would then get a crash on first form evaluation, does that sound correct? I think that's ok but want to make sure I understand the implications. If that's all sounding right then this comment can be removed.

Reply:
Hi @lognaturel 

Updated to reflect the changes in the comments